gomap/gomap.gobra

```diff
@@ -78,44 +60,136 @@ pure func IsEmpty(m map[int]int) (result bool) {
 	return len(m) == 0
 }
 
-// skip dicts.Get: option is ghost (results in Logic error: ghost classification on unsafe node)
-// skip Restrict, RemoveKeys, Remove: requires creation of new map
+// Retrieves the value associated with the key, if present, as an option.
+ghost
+requires acc(m, _)
+decreases
+pure func GetOption(m map[int]int, k int) (result option[int]) {
+	return dicts.Get(ToDict(m), k)
+}
 
 // True iff k maps to the same value or is not in the domains of m1 and m2.
-requires acc(m1, 1/2)
-requires acc(m2, 1/2)
-ensures acc(m1, 1/2)
-ensures acc(m2, 1/2)
+ghost
+requires acc(m1, _)
+requires acc(m2, _)
 ensures result == dicts.IsEqualOnKey(ToDict(m1), ToDict(m2), k)
-func IsEqualOnKey(m1, m2 map[int]int, k int) (result bool) {
-	// QUES If I don't factor out Exists like this, I cannot prove the postcondition. Why?
-	// Calling exists multiple times shouldn't change anything, as we only get read-only access
-	// to heaps and nothing else.
-	exists1 := Exists(m1, k)
-	exists2 := Exists(m2, k)
-	return !(exists1 || exists2) ||
-		(exists1 && exists2 && Get(m1, k) == Get(m2, k))
+decreases
+pure func IsEqualOnKey(m1, m2 map[int]int, k int) (result bool) {
+	return !(k in domain(m1) || k in domain(m2)) ||
+		(k in domain(m1) && k in domain(m2) && m1[k] == m2[k])
 }
 
 // True iff m1 is a subset of m2.
-requires acc(m1, 1/2)
-requires acc(m2, 1/2)
-ensures acc(m1, 1/2)
-ensures acc(m2, 1/2)
+ghost
+requires acc(m1, _)
+requires acc(m2, _)
 ensures result == dicts.IsSubset(ToDict(m1), ToDict(m2))
-func IsSubset(m1, m2 map[int]int) (result bool) {
-	invariant acc(m1, 1/2)
-	invariant acc(m2, 1/2)
-	invariant visited subset domain(ToDict(m2))
-	invariant forall k int :: k in visited ==> dicts.IsEqualOnKey(ToDict(m1), ToDict(m2), k)
-	for k := range m1 with visited {
-		assert acc(m1, 1/2)
-		v1 := Get(m1, k)
-		v2, ok := GetExists(m2, k)
-		if v1 != v2 || !ok {
-			return false
-		}
-	}
-	
-	return true
+decreases
+pure func IsSubset(m1, m2 map[int]int) (result bool) {
+	return domain(m1) subset domain(m2) &&
+		forall k int :: {IsEqualOnKey(m1, m2, k)} {k in domain(m1)} (k in domain(m1)) ==>
+		IsEqualOnKey(m1, m2, k)
+}
+
+// Maps are disjoint iff their domains are disjoint.
+ghost
+requires acc(m1, _)
+requires acc(m2, _)
+ensures result == dicts.AreDisjoint(ToDict(m1), ToDict(m2))
+decreases
+pure func AreDisjoint(m1, m2 map[int]int) (result bool) {
+	return sets.AreDisjoint(domain(m1), domain(m2))
+}
+
+// True iff a map is injective.
+ghost
+requires acc(m, _)
+// QUES This postcondition (and many of the following) don't verify either. Since
+// Agree also does not verify even though it is not opaque, I would guess
+// it has something to do with forall/triggers (even though IsSubset verifies just fine)
+ensures result == dicts.IsInjective(ToDict(m))
```

Comment:
Have you tried asserting the postcondition of `ToDict` immediately after the call to `ToDict(m)`?